File: demo-android/app/src/main/java/com/example/reports/Report.kt
package com.example.reports

data class Row(val label: String, val value: Int)

// A Report holds rows and renders them in a chosen format.
class Report(val title: String, val rows: List<Row>) {

    fun render(format: String = "text"): String {
        val formatter: ReportFormatter = if (format == "csv") CsvFormatter() else TextFormatter()
        return formatter.format(title, rows)
    }
}

File: validation/demo-py/report.py
"""A tiny report renderer. `render` is the text default; `render_as` selects a
formatter by name."""

from formatter import TextFormatter
from csvfmt import CsvFormatter

_FORMATTERS = {"text": TextFormatter, "csv": CsvFormatter}


def render(title, rows):
    return TextFormatter().format(title, rows)


def render_as(fmt, title, rows):
    if fmt not in _FORMATTERS:
        raise ValueError(f"Unknown format: {fmt}")
    return _FORMATTERS[fmt]().format(title, rows)
